File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.py
from typing import List

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = []

        for num in nums:
            l,r = 0, len(dp)

            while l < r:
                m = (r+l)//2

                if dp[m] < num:
                    l = m+1
                else:
                    r = m
            
            if l == len(dp):
                dp.append(num)
            else:
                dp[l] = num

        return len(dp)

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (sum%2 == 1) {return false;}
        vector<bool> dp(sum + 1, false);
        dp[0] = true;

        for (int num: nums) {
            for (int i = sum / 2 + 1; i >= 0; --i) {
                if (dp[i]) {
                    dp[i + num] = true;
                    if (i + num == sum / 2)  {
                        return true;
                    }
                }
            }
            dp[num] = true;
        }

        return false;
    }
};
